### ISSUE TITLE
Fortran 2003 and conditionals

### ISSUE THREAD

**Issue by cmbant** [textmate/fortran.tmbundle#19]
"optional", "allocatable", "class", "Type, extends", and other elements of Fortran 2003 code are not highlighted consistently. Also #ifdef, #endif are not highlighted clearly differently.
 e.g. see

https://github.com/cmbant/forutils/blob/master/ObjectLists.f90
https://github.com/cmbant/forutils/blob/master/MpiUtils.f90

**Comment by cmbant** [comment]
PS. I notice in previous bug reports you ask for a reference for the language structure. Here's the exhaustive specification of the language:

http://www.j3-fortran.org/doc/year/10/10-007.pdf

**Comment by infininight** [comment]
I believe the `optional` and `allocatable` cases have been corrected, if there are any I missed give line numbers.

Class and extends are not constructs currently supported by the grammar, so they are not intended to be matched at all currently. (Neither is `Type, abstract` which I note from your example files.)

Same with the conditional lines, though I do not see those mentioned at all in the document linked.

**Comment by cmbant** [comment]
The MpiUtils.f90 file has #ifdef MPI etc, which are currently highlighted the same way a fortran code.

**Comment by infininight** [comment]
Yes I see them there I meant I don't see them mentioned in the spec document you linked.

**Comment by cmbant** [comment]
Ah, Fortran just uses the standard C preprocessor directives.

**Comment by wusunlab** [comment]
To add to this, `pure` and `elemental` procedures are also not highlighted consistently.  If you have some thing like `pure function ...` it will highlight; but if you have `pure real(8) function ...` it won't highlight it properly.
